### PR TITLE
MILESTONE: Merging `development` into `main` 🎉 

### LINKED ISSUE #220: [DEVELOPMENT] Utilise `edge_attr` in graph convolutions
Source information [here](https://github.com/pyg-team/pytorch_geometric/issues/668)

### LINKED ISSUE #222: [DEVELOPMENT] Visualise the whole graph inference after each epoch

### LINKED ISSUE #244: [DEVELOPMENT] Generalise `GCN` for `GNNClassifier` object for training
Relates to #100

### LINKED ISSUE #247: [DEVELOPMENT] How to set a global config with different font attributes in matplotlib?
I'm looking for an equivalent global config as the logger has:

https://github.com/alan-turing-institute/grace/blob/bd91bdffe0aa3ad3ed6bab0f97cb1d90f0b37ad2/grace/logger.py#L4-L8

Something in the line of this snippet, which would be defined once (probably in `run.py`) and inherited all over the pipeline:
```
font = {'family' : 'normal',
        'weight' : 'bold',
        'size'   : 22}

matplotlib.rc('font', **font)
```
Any suggestions?

### LINKED ISSUE #254: [FEATURE REQUEST] Consider using `.yaml` files for hyperparameter config
A bit more elegant to read 
```sh
metrics:
- accuracy
- confusion_matrix
```
than
`"metrics": "['accuracy', 'confusion_matrix']",`

### LINKED ISSUE #255: [DEVELOPMENT] Transfer the `PREDICTION` probabilities into a dictionary
In this way, avoid indexing confusion & improve graph legibility:

https://github.com/alan-turing-institute/grace/blob/5dd7104ea37a3de5043107f0bb3c041b3f035aab/grace/base.py#L104-L105

_Suggestion by @crangelsmith_ 🚀 

The implementation is actually more elegant, using a `Prediction` dataclass:

https://github.com/alan-turing-institute/grace/blob/7fcc1df4465baa5755c079f3ac65753d17cdd796/grace/base.py#L38-L89

### LINKED ISSUE #257: [FEATURE REQUEST] Log F1-score metric to Tensorboard log
**Please describe the GRACE feature 🪄 you'd like to add**, including why this is useful & the type(s) of data you're working on.

### LINKED ISSUE #258: [FEATURE REQUEST] Define a configurable dictonary of colormaps throughout the pipeline

### LINKED ISSUE #261: [FEATURE REQUEST] Implement optimiser weight decay
https://github.com/alan-turing-institute/grace/blob/bd91bdffe0aa3ad3ed6bab0f97cb1d90f0b37ad2/grace/training/train.py#L77-L82

### LINKED ISSUE #263: [FEATURE REQUEST] Pool all colour selections into `styling.py` script

### LINKED ISSUE #281: [DEVELOPMENT] Allow specifying variable number of graph conv & dense linear layers

### LINKED ISSUE #282: [FEATURE REQUEST] Output the summary of the model architecture for completeness

### LINKED ISSUE #284: [DEVELOPMENT] Store edge properties in `Data` object for training
**Describe the issue here**

### LINKED ISSUE #291: Add more sophisticated classifier model options
Remove the placeholder for other classifier model options. As soon as I'm done with edge property calculation, I can extend the model options to other architectures (I'm unsure how to design this properly tho, to avoid unnecessary code duplication).

_Originally posted by @KristinaUlicna in https://github.com/alan-turing-institute/grace/pull/283#discussion_r1345530374_

### LINKED ISSUE #294: [FEATURE REQUEST] Consider visualising the edge properties in the dataset
To make sure your properties are of reasonable ranges / bounds / distributions, consider plotting out their histograms.

### LINKED ISSUE #296: [BUG] Model architecture summary doesn't save up GNN layers by name

### LINKED ISSUE #297: [DEVELOPMENT] Allow feature / property extraction to be computed at run-time
2 issues here:
- if node features / edge properties graph attributes are not present, compute them & optionally store in the files.
- if augmentations are configured, perform on the run computations.

### LINKED ISSUE #299: [DEVELOPMENT] Write proper documentation for `EdgePropertyCruncher`
**Describe the issue here**

### LINKED ISSUE #303: [DEVELOPMENT] Compute edge properties
https://github.com/alan-turing-institute/grace/blob/b59c4d5df6bdd4c41b76cfefa9e364816a4469b1/grace/base.py#L93-L160

### LINKED ISSUE #304: [DEVELOPMENT] Define an `EdgePropertyCruncher`
Implementation here:

https://github.com/alan-turing-institute/grace/blob/b59c4d5df6bdd4c41b76cfefa9e364816a4469b1/grace/models/property_cruncher.py#L7-L19

### LINKED ISSUE #310: Clarify which properties are used in the training
It might be useful to be clear about which properties are actually used in the training, we know the `raw` properties can't be passed directly to the network because it would memorise it, so lets make sure the user knows that.

_Originally posted by @crangelsmith in https://github.com/alan-turing-institute/grace/pull/307#discussion_r1358052268_

### LINKED ISSUE #315: [BUG] Undefined `roc_auc_score` if only one class present in `y_true`
Tests fail due to this.

### LINKED ISSUE #318: [FEATURE REQUEST] Separate inference into a stand-alone module
Make inference reusable by other scripts parts of the pipeline.

### LINKED ISSUE #330: [DEVELOPMENT] Concatenate Laplacian-multiplied node features

### PR DESCRIPTION
**Pull request by KristinaUlicna**
## PR contribution summary

**Why is this PR useful / good for?** *Please describe the problem(s) you're trying to address.*
* Major merger between `development` and `main`

### List of proposed changes / linked issues & discussions

Among the already closed issues, this PR will close:

- [x] ✅ Resolves #220
- [x] ✅ Resolves #222
- [x] ✅ Resolves #244
- [x] ✅ Resolves #247
- [x] ✅ Resolves #254
- [x] ✅ Resolves #255
- [x] ✅ Resolves #257
- [x] ✅ Resolves #258
- [x] ✅ Resolves #261
- [x] ✅ Resolves #263
- [x] ✅ Resolves #281
- [x] ✅ Resolves #282
- [x] ✅ Resolves #284
- [x] ✅ Resolves #291
- [x] ✅ Resolves #294
- [x] ✅ Resolves #296
- [x] ✅ Resolves #297
- [x] ✅ Resolves #284
- [x] ✅ Resolves #299
- [x] ✅ Resolves #303
- [x] ✅ Resolves #304
- [x] ✅ Resolves #310
- [x] ✅ Resolves #315
- [x] ✅ Resolves #318
- [x] ✅ Resolves #330

### What should a reviewer concentrate their feedback on?
Major merger, all changes have been reviewed when merged to `development`.

#### What type of PR is this? (check all applicable)

+ 🪄 Feature
+ 🐛 Bug fix
+ 📝 Documentation / update README.md
+ #️⃣ Documentation / code annotation
+ 🧑‍💻 Code refactor / style
+ 🔥 Performance Improvements
+ ✅ Test

#### Added tests?

+ 👍 yes
+ 🙅 no, because they aren't needed
+ 🙋 no, because I need some help

---

## PR review summary

Describe what this PR does & **how you reviewed** the individual items, where needed:

* <TEXT>

*Some helper checks to tick off:*

* Focus on image annotation
* Focus on model training
* Could any optimization be applied?
* Is there any redundant code?
* Are there any spelling errors?

**In conclusion**, after my review, I'd like to:

* 🙋 ask some clarifying questions
* 🙅 suggest some specific changes

### CODE REVIEW

**Review by marjanfamili — APPROVED** [review]
This branch runs without error for the following config file.  

[config_hyperparams.txt](https://github.com/alan-turing-institute/grace/files/13333408/config_hyperparams.txt)
